crates/red_knot_python_semantic/resources/mdtest/generics/pep695.md

```diff
@@ -48,4 +48,120 @@ class C[T]:
+def unbounded_unconstrained[T, U](t: T, u: U) -> None:
+    static_assert(is_assignable_to(T, T))
+    static_assert(is_assignable_to(U, U))
+    static_assert(not is_assignable_to(T, U))
+    static_assert(not is_assignable_to(U, T))
+
+    static_assert(not is_subtype_of(T, T))
+    static_assert(not is_subtype_of(U, U))
+    static_assert(not is_subtype_of(T, U))
+    static_assert(not is_subtype_of(U, T))
+```
+
+A bounded typevar is assignable to its bound, but the bound is not assignable to the typevar (since
+the typevar might be specialized to a smaller type). (TODO: Unless the bound is final, in which case
+the final class is also assignable to the typevar.)
+
+```py
+from typing_extensions import final
+
+def bounded[T: int](t: T) -> None:
+    static_assert(is_assignable_to(T, int))
+    static_assert(not is_assignable_to(int, T))
+
+    static_assert(not is_subtype_of(T, int))
+    static_assert(not is_subtype_of(int, T))
+
+@final
+class FinalClass: ...
+
+def bounded_final[T: FinalClass](t: T) -> None:
+    static_assert(is_assignable_to(T, FinalClass))
+    # TODO: is_assignable_to
+    static_assert(not is_assignable_to(FinalClass, T))
+
+    static_assert(not is_subtype_of(T, FinalClass))
+    static_assert(not is_subtype_of(FinalClass, T))
+```
+
+Two distinct typevars are not assignable to each other, even if they have the same bounds, since
+there is (still) no guarantee that they will be specialized to the same type. (TODO: Unless the
+bound is final, in which case we _can_ assume that the two typevars must always be specialized to
+that final class.)
+
+```py
+def two_bounded[T: int, U: int](t: T, u: U) -> None:
+    static_assert(not is_assignable_to(T, U))
+    static_assert(not is_assignable_to(U, T))
+
+    static_assert(not is_subtype_of(T, U))
+    static_assert(not is_subtype_of(U, T))
+
+def two_final_bounded[T: FinalClass, U: FinalClass](t: T, u: U) -> None:
+    # TODO: is_assignable_to
+    static_assert(not is_assignable_to(T, U))
+    # TODO: is_assignable_to
+    static_assert(not is_assignable_to(U, T))
+
+    static_assert(not is_subtype_of(T, U))
+    static_assert(not is_subtype_of(U, T))
+```
+
+A constrained typevar is assignable to the union of its constraints, but not to any of the
+constraints individually. None of the constraints are assignable to the typevar.
+
+```py
+def constrained[T: (int, str)](t: T) -> None:
+    static_assert(not is_assignable_to(T, int))
+    static_assert(not is_assignable_to(T, str))
+    static_assert(is_assignable_to(T, int | str))
+    static_assert(not is_assignable_to(int, T))
+    static_assert(not is_assignable_to(str, T))
+    static_assert(not is_assignable_to(int | str, T))
+
+    static_assert(not is_subtype_of(T, int))
+    static_assert(not is_subtype_of(T, str))
+    static_assert(not is_subtype_of(T, int | str))
+    static_assert(not is_subtype_of(int, T))
+    static_assert(not is_subtype_of(str, T))
+    static_assert(not is_subtype_of(int | str, T))
+```
```

Comment:
Would it make sense to add a test like the following? I could imagine that code like this gets written if a generic function needs special handling for a particular special case?
```py
class P: ...
class Q: ...

def constrained[T: (P, Q)](t: T) -> None:
    if isinstance(t, P):
        p: P = t  # this works already on your branch
    else:
        q: Q = t  # this yields an error, but should be fine(?)
```
Alternatively, we could also use `reveal_type(t) in those branches and assert on the narrowed types of `t`. We currently reveal `T & P` and `T & ~P`, which might be simplified to `P` and `Q`?

Reply:
Yes, good idea.  I hadn't touched the narrowing code, but should.

Reply:
Looking at this more, I think this is because of how I've implemented `is_subtype_of` for typevars. Since they can be specialized to `Any`, I'm treating them as not fully static, so they don't participate in subtyping at all. If we loosened that a bit for constrained types, I think the existing intersection builder logic would simplify these as you suggest.